### PR TITLE
BufferTimeout with fair backpressure rework

### LINKED ISSUE #3531: BufferTimeout or FluxFlatmapIterable fails to discard all elements
```
OnDiscardShouldNotLeakTest > Populated queue racing Cancel/overflow Error > Populated queue racing Cancel/overflow Error [73] for bufferTimeout FAILED
    java.lang.AssertionError: 
    Expecting all elements of:
      [Tracked{ id=0 released=true },
        Tracked{ id=1 released=true },
        Tracked{ id=2 released=true },
        Tracked{ id=3 released=true },
        Tracked{ id=4 released=true },
        Tracked{ id=5 released=true },
        Tracked{ id=6 released=true },
        Tracked{ id=7 released=true },
        Tracked{ id=8 released=true },
        Tracked{ id=9 released=true },
        Tracked{ id=10 released=true },
        Tracked{ id=11 released=true },
        Tracked{ id=12 released=true },
        Tracked{ id=13 released=true },
        Tracked{ id=14 released=true },
        Tracked{ id=15 released=true },
        Tracked{ id=16 released=true },
        Tracked{ id=17 released=true },
        Tracked{ id=18 released=true },
        Tracked{ id=19 released=true },
        Tracked{ id=20 released=true },
        Tracked{ id=21 released=true },
        Tracked{ id=22 released=true },
        Tracked{ id=23 released=true },
        Tracked{ id=24 released=true },
        Tracked{ id=25 released=true },
        Tracked{ id=26 released=true },
        Tracked{ id=27 released=true },
        Tracked{ id=28 released=true },
        Tracked{ id=29 released=true },
        Tracked{ id=30 released=true },
        Tracked{ id=31 released=true },
        Tracked{ id=32 released=true },
        Tracked{ id=33 released=true },
        Tracked{ id=34 released=true },
        Tracked{ id=35 released=true },
        Tracked{ id=36 released=true },
        Tracked{ id=37 released=true },
        Tracked{ id=38 released=true },
        Tracked{ id=39 released=true },
        Tracked{ id=40 released=true },
        Tracked{ id=41 released=true },
        Tracked{ id=42 released=true },
        Tracked{ id=43 released=true },
        Tracked{ id=44 released=true },
        Tracked{ id=45 released=true },
        Tracked{ id=46 released=true },
        Tracked{ id=47 released=true },
        Tracked{ id=48 released=true },
        Tracked{ id=49 released=true },
        Tracked{ id=50 released=true },
        Tracked{ id=51 released=true },
        Tracked{ id=52 released=true },
        Tracked{ id=53 released=true },
        Tracked{ id=54 released=true },
        Tracked{ id=55 released=true },
        Tracked{ id=56 released=true },
        Tracked{ id=57 released=true },
        Tracked{ id=58 released=true },
        Tracked{ id=59 released=true },
        Tracked{ id=60 released=true },
        Tracked{ id=61 released=true },
        Tracked{ id=62 released=true },
        Tracked{ id=63 released=true },
        Tracked{ id=64 released=true },
        Tracked{ id=65 released=true },
        Tracked{ id=66 released=true },
        Tracked{ id=67 released=true },
        Tracked{ id=68 released=true },
        Tracked{ id=69 released=true },
        Tracked{ id=70 released=true },
        Tracked{ id=71 released=true },
        Tracked{ id=72 released=true },
        Tracked{ id=73 released=true },
        Tracked{ id=74 released=true },
        Tracked{ id=75 released=true },
        Tracked{ id=76 released=true },
        Tracked{ id=77 released=true },
        Tracked{ id=78 released=true },
        Tracked{ id=79 released=true },
        Tracked{ id=80 released=true },
        Tracked{ id=81 released=true },
        Tracked{ id=82 released=true },
        Tracked{ id=83 released=true },
        Tracked{ id=84 released=true },
        Tracked{ id=85 released=true },
        Tracked{ id=86 released=true },
        Tracked{ id=87 released=true },
        Tracked{ id=88 released=true },
        Tracked{ id=89 released=true },
        Tracked{ id=90 released=true },
        Tracked{ id=91 released=true },
        Tracked{ id=92 released=true },
        Tracked{ id=93 released=true },
        Tracked{ id=94 released=true },
        Tracked{ id=95 released=true },
        Tracked{ id=96 released=true },
        Tracked{ id=97 released=true },
        Tracked{ id=98 released=true },
        Tracked{ id=99 released=true },
        Tracked{ id=100 released=true },
        Tracked{ id=101 released=true },
        Tracked{ id=102 released=true },
        Tracked{ id=103 released=true },
        Tracked{ id=104 released=true },
        Tracked{ id=105 released=true },
        Tracked{ id=106 released=true },
        Tracked{ id=107 released=true },
        Tracked{ id=108 released=true },
        Tracked{ id=109 released=true },
        Tracked{ id=110 released=true },
        Tracked{ id=111 released=true },
        Tracked{ id=112 released=true },
        Tracked{ id=113 released=true },
        Tracked{ id=114 released=true },
        Tracked{ id=115 released=true },
        Tracked{ id=116 released=true },
        Tracked{ id=117 released=true },
        Tracked{ id=118 released=true },
        Tracked{ id=119 released=true },
        Tracked{ id=120 released=true },
        Tracked{ id=121 released=true },
        Tracked{ id=122 released=true },
        Tracked{ id=123 released=true },
        Tracked{ id=124 released=true },
        Tracked{ id=125 released=true },
        Tracked{ id=126 released=true },
        Tracked{ id=127 released=true },
        Tracked{ id=128 released=true },
        Tracked{ id=129 released=true },
        Tracked{ id=130 released=true },
        Tracked{ id=131 released=true },
        Tracked{ id=132 released=true },
        Tracked{ id=133 released=true },
        Tracked{ id=134 released=true },
        Tracked{ id=135 released=true },
        Tracked{ id=136 released=true },
        Tracked{ id=137 released=true },
        Tracked{ id=138 released=true },
        Tracked{ id=139 released=true },
        Tracked{ id=140 released=true },
        Tracked{ id=141 released=true },
        Tracked{ id=142 released=true },
        Tracked{ id=143 released=true },
        Tracked{ id=144 released=true },
        Tracked{ id=145 released=true },
        Tracked{ id=146 released=true },
        Tracked{ id=147 released=true },
        Tracked{ id=148 released=true },
        Tracked{ id=149 released=true },
        Tracked{ id=150 released=true },
        Tracked{ id=151 released=true },
        Tracked{ id=152 released=true },
        Tracked{ id=153 released=true },
        Tracked{ id=154 released=true },
        Tracked{ id=155 released=true },
        Tracked{ id=156 released=true },
        Tracked{ id=157 released=true },
        Tracked{ id=158 released=true },
        Tracked{ id=159 released=true },
        Tracked{ id=160 released=true },
        Tracked{ id=161 released=true },
        Tracked{ id=162 released=true },
        Tracked{ id=163 released=true },
        Tracked{ id=164 released=true },
        Tracked{ id=165 released=true },
        Tracked{ id=166 released=true },
        Tracked{ id=167 released=true },
        Tracked{ id=168 released=true },
        Tracked{ id=169 released=true },
        Tracked{ id=170 released=true },
        Tracked{ id=171 released=true },
        Tracked{ id=172 released=true },
        Tracked{ id=173 released=true },
        Tracked{ id=174 released=true },
        Tracked{ id=175 released=true },
        Tracked{ id=176 released=true },
        Tracked{ id=177 released=true },
        Tracked{ id=178 released=true },
        Tracked{ id=179 released=true },
        Tracked{ id=180 released=true },
        Tracked{ id=181 released=true },
        Tracked{ id=182 released=true },
        Tracked{ id=183 released=true },
        Tracked{ id=184 released=true },
        Tracked{ id=185 released=true },
        Tracked{ id=186 released=true },
        Tracked{ id=187 released=true },
        Tracked{ id=188 released=true },
        Tracked{ id=189 released=true },
        Tracked{ id=190 released=true },
        Tracked{ id=191 released=true },
        Tracked{ id=192 released=true },
        Tracked{ id=193 released=true },
        Tracked{ id=194 released=true },
        Tracked{ id=195 released=true },
        Tracked{ id=196 released=true },
        Tracked{ id=197 released=true },
        Tracked{ id=198 released=true },
        Tracked{ id=199 released=true },
        Tracked{ id=200 released=true },
        Tracked{ id=201 released=true },
        Tracked{ id=202 released=true },
        Tracked{ id=203 released=true },
        Tracked{ id=204 released=true },
        Tracked{ id=205 released=true },
        Tracked{ id=206 released=true },
        Tracked{ id=207 released=true },
        Tracked{ id=208 released=true },
        Tracked{ id=209 released=true },
        Tracked{ id=210 released=true },
        Tracked{ id=211 released=true },
        Tracked{ id=212 released=true },
        Tracked{ id=213 released=true },
        Tracked{ id=214 released=true },
        Tracked{ id=215 released=true },
        Tracked{ id=216 released=true },
        Tracked{ id=217 released=true },
        Tracked{ id=218 released=true },
        Tracked{ id=219 released=true },
        Tracked{ id=220 released=true },
        Tracked{ id=221 released=true },
        Tracked{ id=222 released=true },
        Tracked{ id=223 released=true },
        Tracked{ id=224 released=true },
        Tracked{ id=225 released=true },
        Tracked{ id=226 released=true },
        Tracked{ id=227 released=true },
        Tracked{ id=228 released=true },
        Tracked{ id=229 released=true },
        Tracked{ id=230 released=true },
        Tracked{ id=231 released=true },
        Tracked{ id=232 released=true },
        Tracked{ id=233 released=true },
        Tracked{ id=234 released=true },
        Tracked{ id=235 released=true },
        Tracked{ id=236 released=true },
        Tracked{ id=237 released=true },
        Tracked{ id=238 released=true },
        Tracked{ id=239 released=true },
        Tracked{ id=240 released=true },
        Tracked{ id=241 released=true },
        Tracked{ id=242 released=true },
        Tracked{ id=243 released=true },
        Tracked{ id=244 released=true },
        Tracked{ id=245 released=true },
        Tracked{ id=246 released=true },
        Tracked{ id=247 released=true },
        Tracked{ id=248 released=true },
        Tracked{ id=249 released=true },
        Tracked{ id=250 released=true },
        Tracked{ id=251 released=true },
        Tracked{ id=252 released=true },
        Tracked{ id=253 released=true },
        Tracked{ id=254 released=true },
        Tracked{ id=255 released=false },
        Tracked{ id=256 released=true },
        Tracked{ id=257 released=true }]
    to match 'isReleased' predicate but this element did not:
      Tracked{ id=255 released=false }
        at reactor.test.MemoryUtils$OffHeapDetector.assertNoLeaks(MemoryUtils.java:144)
        at reactor.core.publisher.OnDiscardShouldNotLeakTest.ensureNoLeaksPopulatedQueueAndRacingCancelAndOverflowError(OnDiscardShouldNotLeakTest.java:586)

OnDiscardShouldNotLeakTest > Populated queue racing Cancel/overflow Error > Populated queue racing Cancel/overflow Error [75] for bufferTimeout (fused) FAILED
    java.lang.AssertionError: 
    Expecting all elements of:
      [Tracked{ id=0 released=true },
        Tracked{ id=1 released=true },
        Tracked{ id=2 released=true },
        Tracked{ id=3 released=true },
        Tracked{ id=4 released=true },
        Tracked{ id=5 released=true },
        Tracked{ id=6 released=true },
        Tracked{ id=7 released=true },
        Tracked{ id=8 released=true },
        Tracked{ id=9 released=true },
        Tracked{ id=10 released=true },
        Tracked{ id=11 released=true },
        Tracked{ id=12 released=true },
        Tracked{ id=13 released=true },
        Tracked{ id=14 released=true },
        Tracked{ id=15 released=true },
        Tracked{ id=16 released=true },
        Tracked{ id=17 released=true },
        Tracked{ id=18 released=true },
        Tracked{ id=19 released=true },
        Tracked{ id=20 released=true },
        Tracked{ id=21 released=true },
        Tracked{ id=22 released=true },
        Tracked{ id=23 released=true },
        Tracked{ id=24 released=true },
        Tracked{ id=25 released=true },
        Tracked{ id=26 released=true },
        Tracked{ id=27 released=true },
        Tracked{ id=28 released=true },
        Tracked{ id=29 released=true },
        Tracked{ id=30 released=true },
        Tracked{ id=31 released=true },
        Tracked{ id=32 released=true },
        Tracked{ id=33 released=true },
        Tracked{ id=34 released=true },
        Tracked{ id=35 released=true },
        Tracked{ id=36 released=true },
        Tracked{ id=37 released=true },
        Tracked{ id=38 released=true },
        Tracked{ id=39 released=true },
        Tracked{ id=40 released=true },
        Tracked{ id=41 released=true },
        Tracked{ id=42 released=true },
        Tracked{ id=43 released=true },
        Tracked{ id=44 released=true },
        Tracked{ id=45 released=true },
        Tracked{ id=46 released=true },
        Tracked{ id=47 released=true },
        Tracked{ id=48 released=true },
        Tracked{ id=49 released=true },
        Tracked{ id=50 released=true },
        Tracked{ id=51 released=true },
        Tracked{ id=52 released=true },
        Tracked{ id=53 released=true },
        Tracked{ id=54 released=true },
        Tracked{ id=55 released=true },
        Tracked{ id=56 released=true },
        Tracked{ id=57 released=true },
        Tracked{ id=58 released=true },
        Tracked{ id=59 released=true },
        Tracked{ id=60 released=true },
        Tracked{ id=61 released=true },
        Tracked{ id=62 released=true },
        Tracked{ id=63 released=true },
        Tracked{ id=64 released=true },
        Tracked{ id=65 released=true },
        Tracked{ id=66 released=true },
        Tracked{ id=67 released=true },
        Tracked{ id=68 released=true },
        Tracked{ id=69 released=true },
        Tracked{ id=70 released=true },
        Tracked{ id=71 released=true },
        Tracked{ id=72 released=true },
        Tracked{ id=73 released=true },
        Tracked{ id=74 released=true },
        Tracked{ id=75 released=true },
        Tracked{ id=76 released=true },
        Tracked{ id=77 released=true },
        Tracked{ id=78 released=true },
        Tracked{ id=79 released=true },
        Tracked{ id=80 released=true },
        Tracked{ id=81 released=true },
        Tracked{ id=82 released=true },
        Tracked{ id=83 released=true },
        Tracked{ id=84 released=true },
        Tracked{ id=85 released=true },
        Tracked{ id=86 released=true },
        Tracked{ id=87 released=true },
        Tracked{ id=88 released=true },
        Tracked{ id=89 released=true },
        Tracked{ id=90 released=true },
        Tracked{ id=91 released=true },
        Tracked{ id=92 released=true },
        Tracked{ id=93 released=true },
        Tracked{ id=94 released=true },
        Tracked{ id=95 released=true },
        Tracked{ id=96 released=true },
        Tracked{ id=97 released=true },
        Tracked{ id=98 released=true },
        Tracked{ id=99 released=true },
        Tracked{ id=100 released=true },
        Tracked{ id=101 released=true },
        Tracked{ id=102 released=true },
        Tracked{ id=103 released=true },
        Tracked{ id=104 released=true },
        Tracked{ id=105 released=true },
        Tracked{ id=106 released=true },
        Tracked{ id=107 released=true },
        Tracked{ id=108 released=true },
        Tracked{ id=109 released=true },
        Tracked{ id=110 released=true },
        Tracked{ id=111 released=true },
        Tracked{ id=112 released=true },
        Tracked{ id=113 released=true },
        Tracked{ id=114 released=true },
        Tracked{ id=115 released=true },
        Tracked{ id=116 released=true },
        Tracked{ id=117 released=true },
        Tracked{ id=118 released=true },
        Tracked{ id=119 released=true },
        Tracked{ id=120 released=true },
        Tracked{ id=121 released=true },
        Tracked{ id=122 released=true },
        Tracked{ id=123 released=true },
        Tracked{ id=124 released=true },
        Tracked{ id=125 released=true },
        Tracked{ id=126 released=true },
        Tracked{ id=127 released=true },
        Tracked{ id=128 released=true },
        Tracked{ id=129 released=true },
        Tracked{ id=130 released=true },
        Tracked{ id=131 released=true },
        Tracked{ id=132 released=true },
        Tracked{ id=133 released=true },
        Tracked{ id=134 released=true },
        Tracked{ id=135 released=true },
        Tracked{ id=136 released=true },
        Tracked{ id=137 released=true },
        Tracked{ id=138 released=true },
        Tracked{ id=139 released=true },
        Tracked{ id=140 released=true },
        Tracked{ id=141 released=true },
        Tracked{ id=142 released=true },
        Tracked{ id=143 released=true },
        Tracked{ id=144 released=true },
        Tracked{ id=145 released=true },
        Tracked{ id=146 released=true },
        Tracked{ id=147 released=true },
        Tracked{ id=148 released=true },
        Tracked{ id=149 released=true },
        Tracked{ id=150 released=true },
        Tracked{ id=151 released=true },
        Tracked{ id=152 released=true },
        Tracked{ id=153 released=true },
        Tracked{ id=154 released=true },
        Tracked{ id=155 released=true },
        Tracked{ id=156 released=true },
        Tracked{ id=157 released=true },
        Tracked{ id=158 released=true },
        Tracked{ id=159 released=true },
        Tracked{ id=160 released=true },
        Tracked{ id=161 released=true },
        Tracked{ id=162 released=true },
        Tracked{ id=163 released=true },
        Tracked{ id=164 released=true },
        Tracked{ id=165 released=true },
        Tracked{ id=166 released=true },
        Tracked{ id=167 released=true },
        Tracked{ id=168 released=true },
        Tracked{ id=169 released=true },
        Tracked{ id=170 released=true },
        Tracked{ id=171 released=true },
        Tracked{ id=172 released=true },
        Tracked{ id=173 released=true },
        Tracked{ id=174 released=true },
        Tracked{ id=175 released=true },
        Tracked{ id=176 released=true },
        Tracked{ id=177 released=true },
        Tracked{ id=178 released=true },
        Tracked{ id=179 released=true },
        Tracked{ id=180 released=true },
        Tracked{ id=181 released=true },
        Tracked{ id=182 released=true },
        Tracked{ id=183 released=true },
        Tracked{ id=184 released=true },
        Tracked{ id=185 released=true },
        Tracked{ id=186 released=true },
        Tracked{ id=187 released=true },
        Tracked{ id=188 released=true },
        Tracked{ id=189 released=true },
        Tracked{ id=190 released=true },
        Tracked{ id=191 released=true },
        Tracked{ id=192 released=true },
        Tracked{ id=193 released=true },
        Tracked{ id=194 released=true },
        Tracked{ id=195 released=true },
        Tracked{ id=196 released=true },
        Tracked{ id=197 released=true },
        Tracked{ id=198 released=true },
        Tracked{ id=199 released=true },
        Tracked{ id=200 released=true },
        Tracked{ id=201 released=true },
        Tracked{ id=202 released=true },
        Tracked{ id=203 released=true },
        Tracked{ id=204 released=true },
        Tracked{ id=205 released=true },
        Tracked{ id=206 released=true },
        Tracked{ id=207 released=true },
        Tracked{ id=208 released=true },
        Tracked{ id=209 released=true },
        Tracked{ id=210 released=true },
        Tracked{ id=211 released=true },
        Tracked{ id=212 released=true },
        Tracked{ id=213 released=true },
        Tracked{ id=214 released=true },
        Tracked{ id=215 released=true },
        Tracked{ id=216 released=true },
        Tracked{ id=217 released=true },
        Tracked{ id=218 released=true },
        Tracked{ id=219 released=true },
        Tracked{ id=220 released=true },
        Tracked{ id=221 released=true },
        Tracked{ id=222 released=true },
        Tracked{ id=223 released=true },
        Tracked{ id=224 released=true },
        Tracked{ id=225 released=true },
        Tracked{ id=226 released=true },
        Tracked{ id=227 released=true },
        Tracked{ id=228 released=true },
        Tracked{ id=229 released=true },
        Tracked{ id=230 released=true },
        Tracked{ id=231 released=true },
        Tracked{ id=232 released=true },
        Tracked{ id=233 released=true },
        Tracked{ id=234 released=true },
        Tracked{ id=235 released=true },
        Tracked{ id=236 released=true },
        Tracked{ id=237 released=true },
        Tracked{ id=238 released=true },
        Tracked{ id=239 released=true },
        Tracked{ id=240 released=true },
        Tracked{ id=241 released=true },
        Tracked{ id=242 released=true },
        Tracked{ id=243 released=true },
        Tracked{ id=244 released=true },
        Tracked{ id=245 released=true },
        Tracked{ id=246 released=true },
        Tracked{ id=247 released=true },
        Tracked{ id=248 released=true },
        Tracked{ id=249 released=true },
        Tracked{ id=250 released=true },
        Tracked{ id=251 released=true },
        Tracked{ id=252 released=true },
        Tracked{ id=253 released=true },
        Tracked{ id=254 released=true },
        Tracked{ id=255 released=false },
        Tracked{ id=256 released=true },
        Tracked{ id=257 released=true }]
    to match 'isReleased' predicate but this element did not:
      Tracked{ id=255 released=false }
        at reactor.test.MemoryUtils$OffHeapDetector.assertNoLeaks(MemoryUtils.java:144)
        at reactor.core.publisher.OnDiscardShouldNotLeakTest.ensureNoLeaksPopulatedQueueAndRacingCancelAndOverflowError(OnDiscardShouldNotLeakTest.java:586)
```

### PR DESCRIPTION
**Pull request by chemicL**
The fair backpressure variant of the `bufferTimeout` operator has been reworked to use a state machine with a minimum number of volatile variables eliminating potential data races, such as skipping the delivery when `onNext` and timeout happen concurrently or cancellation happens while `onNext` is delivered, etc.

Resolves #3531

### CODE REVIEW

**Review by OlegDokuka — CHANGES_REQUESTED** [review]
The issue sits deeper and as I suggested first requires complete rework. 

To reproduce the issue try to modify discard setup as follows: 

```java
@Tag("slow")
public class OnDiscardShouldNotLeakTest {

	private static final int NB_ITERATIONS = 100_000;
	// add DiscardScenarios here to test more operators
	private static final DiscardScenario[] SCENARIOS = new DiscardScenario[] {
			DiscardScenario.fluxSource("bufferTimeout", f -> f.bufferTimeout(2, Duration.ofNanos(1), true).flatMapIterable(Function.identity())),
	};
       ...
  ```
  
  the problem appears when multiple threads comes into play. If you have race between scheduled timeout task trying to flash the window and cancellation ,then some elements might be undischarged

**Comment by OlegDokuka** [comment]
Although the above could be false negative, since the async task can be unawaited properly. Can you please doublecheck @chemicL

**Comment by OlegDokuka** [comment]
I just remembered that it was one of the reasons why all the discard tests are in stress tests, therefore it could make sense to port part of them for bufferTimeout

**Comment by chemicL** [comment]
@OlegDokuka the commit [0334959](https://github.com/reactor/reactor-core/pull/3634/commits/033495922f1a92396def5ea6a240afd2035ba756) adds significant improvements to the test suite that helps catch the racy situations. In the next commit I will add temporary fixes for the identified issues, but afterwards will follow up with a state machine implementation to eliminate the last one with timeout racing with draining.

**Review by OlegDokuka — APPROVED** [review]

**Comment by chemicL** [comment]
For the latest changes, I added a JMH benchmark. The idea is to simply test a non-contended, single threaded case of pulling 100 items and packaging them in 1, 10, and 100-item buffers.

Devising a contended usage scenario with a concurrent, asymmetric JMH benchmark doesn't seem worthwhile as the dominating factor should never be two actors competing but rather one actor winning a potential race and performing the bulk of the work. We can't approach this performance evaluation the same way as a regular queue benchmark with a randezvous scenario of handing over between two threads. In the Queue benchmark case, this operation is the essence of the work being done. In contrast, with reactive-streams it is not the case as there is the idea of work stealing where one actor winning a race (which can contend in a lock-free manner for a short moment) is performing a drain operation and the Queue is a means to an end and we are not evaluating its behaviour under load, but the operator's algorithm itself.

Having said the above, I compared the new backpressured variant with the current (broken) implementation and also with the simpler, yet not-backpressured variant. Here are my results on an M1 MacBook Pro (10-core, 32GB):

```
This PR, fairBackpressure = true

Benchmark                             (bufferSize)  Mode  Cnt     Score     Error  Units
FluxBufferTimeoutBenchmark.oneByOne              1  avgt    5  6259.514 ± 214.272  ns/op
FluxBufferTimeoutBenchmark.oneByOne             10  avgt    5  4418.668 ± 595.467  ns/op
FluxBufferTimeoutBenchmark.oneByOne            100  avgt    5  3938.607 ± 518.551  ns/op
FluxBufferTimeoutBenchmark.unlimited             1  avgt    5  5764.457 ± 399.358  ns/op
FluxBufferTimeoutBenchmark.unlimited            10  avgt    5  4072.673 ± 169.380  ns/op
FluxBufferTimeoutBenchmark.unlimited           100  avgt    5  3797.930 ±  94.629  ns/op

Current 3.5.x, fairBackpressure = true

Benchmark                             (bufferSize)  Mode  Cnt      Score      Error  Units
FluxBufferTimeoutBenchmark.oneByOne              1  avgt    5  23836.157 ± 2138.635  ns/op
FluxBufferTimeoutBenchmark.oneByOne             10  avgt    5   8935.440 ±   43.610  ns/op
FluxBufferTimeoutBenchmark.oneByOne            100  avgt    5   7182.099 ±  129.899  ns/op
FluxBufferTimeoutBenchmark.unlimited             1  avgt    5  18836.570 ± 1281.322  ns/op
FluxBufferTimeoutBenchmark.unlimited            10  avgt    5   8231.299 ±   55.848  ns/op
FluxBufferTimeoutBenchmark.unlimited           100  avgt    5   7197.949 ±  160.759  ns/op

Current 3.5.x, fairBackpressure = false

Benchmark                             (bufferSize)  Mode  Cnt      Score      Error  Units
FluxBufferTimeoutBenchmark.oneByOne              1  avgt    5  18376.986 ±  795.211  ns/op
FluxBufferTimeoutBenchmark.oneByOne             10  avgt    5   3458.388 ±  200.223  ns/op
FluxBufferTimeoutBenchmark.oneByOne            100  avgt    5   2533.851 ±  440.186  ns/op
FluxBufferTimeoutBenchmark.unlimited             1  avgt    5  15716.505 ± 1168.297  ns/op
FluxBufferTimeoutBenchmark.unlimited            10  avgt    5   3040.167 ±  144.305  ns/op
FluxBufferTimeoutBenchmark.unlimited           100  avgt    5   2470.696 ±  375.516  ns/op

```

To summarize the above, the new implementation passes the JCStress tests devised to catch the discovered issues while maintaining a better performance characteristic. The non-prefetching and incapable of respecting backpressure variant is more performant and that's understood as it doesn't need to keep track of much accounting and is a simpler implementation.

File: reactor-core/src/main/java/reactor/core/publisher/StateLogger.java [hunk 1]
```diff
@@ -49,7 +49,7 @@ void log(String instance,
 					formatState(committedState, 64)), new RuntimeException());
 		}
 		else {
-			this.logger.trace(String.format("[%s][%s][%s][%s-%s]",
+			this.logger.trace(String.format("[%s][%s][%s][\n\t%s\n\t%s]",
```
Comment (chemicL) [inline reactor-core/src/main/java/reactor/core/publisher/StateLogger.java:52]: With this change it's much easier to compare the state changes as they are stacked on top of each other so the differences are immediately visible as opposed to a horizontal presentation.

File: reactor-core/src/test/java/reactor/core/publisher/FluxBufferTimeoutFairBackpressureTest.java [hunk 2]
```diff
@@ -365,7 +367,7 @@ public void requestedFromUpstreamShouldNotExceedDownstreamDemand() {
 		            .assertNext(s -> assertThat(s).containsExactly("a"))
 		            .then(() -> assertThat(requestedOutstanding).hasValue(19))
 		            .thenRequest(1)
-		            .then(() -> assertThat(requestedOutstanding).hasValue(20))
+		            .then(() -> assertThat(requestedOutstanding).hasValue(19))
```
Comment (chemicL) [inline reactor-core/src/test/java/reactor/core/publisher/FluxBufferTimeoutFairBackpressureTest.java:370]: This is a behaviour change, yet it is not dramatic or disruptive. Before, a `request` would force `request`ing whatever is remaining from the upstream to fulfill the `prefetch` capacity. Now it will do that only once the `replenishMark` is reached.

**Review by violetagg — APPROVED** [review]
